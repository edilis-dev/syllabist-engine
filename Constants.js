/**
 * @module Constants
 */

/**
 * Key identification types.
 *
 * @see <a href="Expander.html##insert"><code>Expander#insert</code></a>
 */
export const Type = Object.freeze({
  Group: "group",
  Empty: "empty",
  Value: "value",
});

/**
 * Group identification symbols.
 *
 * @see <a href="Compressor.html##object"><code>Compressor#object</code></a>
 * @see <a href="Expander.html##parse"><code>Expander#parse</code></a>
 */
export const Symbol = Object.freeze({
  Combinator: "~",
  Concatenator: ">",
  GroupEnd: "]",
  GroupStart: "[",
  Sibling: "|",
});

/**
 * Supported standardisation formats.
 *
 * @see [Format#Standardise]{@linkcode module:Format.Standardise}
 */
export const Standard = {
  JSON: "json",
  Text: "text",
};

/**
 * Valid character set, including alphabetical characters, hyphens and [Symbols]{@linkcode module:Constants.Symbol}.
 * @see <a href="Expander.html##parse"><code>Expander#parse</code></a>
 */
export const Charset = new RegExp(`[${Object.values(Symbol).join("\\")}a-z-]`);
