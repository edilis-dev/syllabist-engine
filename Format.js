import { Standard } from "./Format.constants.js";

export function Standardise({ data, type = Standard.JSON }) {
  if (!data) {
    throw new TypeError("Empty data");
  }

  switch (type) {
    case Standard.JSON: {
      return JSON.stringify(JSON.parse(data), undefined, 2);
    }
    case Standard.Text: {
      return data.trim();
    }
    default:
      throw new TypeError(`Unhandled type ${type}`);
  }
}
