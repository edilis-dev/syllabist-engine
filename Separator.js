import * as log from "./Log.js";

import {
  BlendTypes,
  CompoundWords,
  DigraphTypes,
  PatternExceptions,
  Prefixes,
  Suffixes,
} from "./Separator.constants.js";
import {
  mergeRepeatedCharacters,
  Groups,
  Patterns,
} from "./Separator.patterns.js";

export class Separator {
  #forward;
  #forwardIndex;
  #lines;
  #minLength = 3;
  #overlap = 1;
  #reverse;
  #reverseIndex;

  constructor(iter) {
    log.info("Constructing");

    this.#lines = iter;
  }

  async separate() {
    log.debug("Separating");

    try {
      let value = [];

      for await (const line of this.#lines) {
        log.info("Starting iteration", {
          line,
        });

        this.#forwardIndex = 0;
        this.#forward = this.forwardIterator(line);

        this.#reverseIndex = line.length;
        this.#reverse = this.reverseIterator(line);

        value = [
          ...value,
          this.#separate({ line })
            .filter(Boolean)
            .join(";")
            .replace(/^;|;$/g, ""),
        ];

        log.debug("Inserting result", {
          value,
        });
      }

      log.info("Result", {
        value,
      });

      return value.join("\n");
    } catch (error) {
      log.error("Error", {
        reason: error.message,
      });

      throw error;
    }
  }

  *forwardIterator(line) {
    if (!line) {
      throw new TypeError("Empty line");
    }

    log.debug("Starting forward iterator", {
      line,
    });

    while (this.#forwardIndex < this.#reverseIndex - this.#overlap) {
      log.debug("Forward iteration", {
        index: this.#forwardIndex,
      });

      yield this.#forwardIndex++;
    }

    return this.#forwardIndex;
  }

  *reverseIterator(line) {
    if (!line) {
      throw new Error("Empty line");
    }

    log.debug("Starting reverse iterator", {
      line,
    });

    while (this.#reverseIndex > this.#forwardIndex - this.#overlap) {
      log.debug("Reseverse iteration", {
        index: this.#reverseIndex,
      });

      yield this.#reverseIndex--;
    }

    return this.#reverseIndex;
  }

  #separate({ cache = {}, line }) {
    const forward = this.#forward.next();
    const backward = this.#reverse.next();

    if (backward.done && forward.done) {
      log.info("Finished parsing prefix and suffix", {
        prefix: cache.prefix,
        suffix: cache.suffix,
      });

      const root = this.#root(cache);

      log.info("Finished parsing root", {
        root,
      });

      return [cache.prefix, root, cache.suffix];
    } else {
      const [start, prefix] = this.#prefix({
        cache,
        index: forward.value,
        line,
      });

      const [end, suffix] = this.#suffix({
        cache,
        index: backward.value,
        line,
      });

      return this.#separate({
        cache: {
          end,
          prefix,
          root: line.slice(start, end),
          start,
          suffix,
        },
        line,
      });
    }
  }

  #prefix({ cache, index, line }) {
    log.debug("Starting prefix parsing", {
      line,
    });

    const prefix = line.slice(0, index);

    if (Prefixes.has(prefix)) {
      log.info("New prefix identified", {
        prefix,
      });

      return [index, prefix];
    } else {
      log.debug("No new prefix identified", {
        index: cache.start,
        prefix: cache.prefix,
      });

      return [cache.start, cache.prefix];
    }
  }

  #root({ root }) {
    log.debug("Starting root parsing", {
      root,
    });

    if (root.length <= this.#minLength) {
      log.info("Word root less than or equal to minimum length", {
        minLenghth: this.#minLength,
        length: root.length,
      });

      return root;
    }

    if (CompoundWords.has(root)) {
      const [a, b] = CompoundWords.get(root);

      log.info("Identified compound word", {
        word: root,
        parts: [a, b],
      });

      return `${this.#root({ root: a })};${this.#root({ root: b })}`;
    }

    // prettier-ignore
    const { BlendSounds, Digraphs, GluedSounds, Quadgraphs, Trigraphs } = Groups;

    const { LE, VCCCCV, VCCCV, VCCV, VCV, VV } = Patterns;

    const tokens = mergeRepeatedCharacters(root);

    log.debug("Tokenised word characters", {
      tokens,
    });

    if (LE.test(tokens)) {
      log.info("Pattern match", {
        pattern: "LE",
      });

      const match = LE.exec(root);

      if (match?.groups) {
        const {
          groups: { beginning, end },
        } = match;

        log.debug("Match groups", {
          beginning,
          end,
        });

        return `${beginning};${this.#root({ root: end })}`;
      }
    } else if (VCCCCV.test(tokens)) {
      log.info("Pattern match", {
        pattern: "VCCCCV",
      });

      const match = VCCCCV.exec(root);

      if (match?.groups) {
        const {
          groups: { beginning, middle, end },
        } = match;

        log.debug("Match groups", {
          beginning,
          middle,
          end,
        });

        // prettier-ignore
        return `${beginning}${middle.at(0)};${middle.at(1)}${middle.at(2)}${middle.at(3)}${end}`;
      }
    } else if (VCCCV.test(tokens)) {
      log.info("Pattern match", { pattern: "VCCCV" });

      const match = VCCCV.exec(root);

      if (match?.groups) {
        const {
          groups: { beginning, middle, end },
        } = match;

        log.debug("Match groups", {
          beginning,
          middle,
          end,
        });

        if (GluedSounds.test(`${beginning.at(-1)}${middle.slice(0, -1)}`)) {
          log.info("Exception identified", {
            type: PatternExceptions.Glued,
            value: `${beginning.at(-1)}${middle.slice(0, -1)}`,
          });

          return `${beginning.slice(0, -1)}${beginning.at(-1)}${middle.at(0)}${middle.at(1)};${this.#root({ root: `${middle.at(2)}${end}` })}`;
        } else {
          log.info("No exception identified");

          return `${beginning}${middle.at(0)};${this.#root({ root: `${middle.slice(1)}${end}` })}`;
        }
      }
    } else if (VCCV.test(tokens)) {
      log.info("Pattern match", {
        pattern: "VCCV",
      });

      const match = VCCV.exec(root);

      if (match?.groups) {
        const {
          groups: { beginning, middle, end },
        } = match;

        log.debug("Match groups", {
          beginning,
          middle,
          end,
        });

        if (BlendSounds.test(middle, BlendTypes.R)) {
          log.info("Exception identified", {
            type: PatternExceptions.Blend.R,
            value: middle,
          });

          return `${beginning};${middle}${this.#root({ root: end })}`;
        } else if (Digraphs.test(middle, DigraphTypes.Consonant)) {
          log.info("Exception identified", {
            type: PatternExceptions.Digraph.Consonant,
            value: middle,
          });

          return `${beginning}${middle};${this.#root({ root: end })}`;
        } else if (Trigraphs.test(root)) {
          log.info("Exception identified", {
            type: PatternExceptions.Trigraph,
            value: root,
          });

          return `${beginning}${middle}${end}`;
        } else {
          log.info("No exception identified");

          return `${beginning}${middle.at(0)};${this.#root({ root: `${middle.at(1)}${end}` })}`;
        }
      }
    } else if (VCV.test(tokens)) {
      log.info("Pattern match", {
        pattern: "VCV",
      });

      const match = VCV.exec(root);

      if (match?.groups) {
        const {
          groups: { beginning, middle, end },
        } = match;

        log.debug("Match groups", {
          beginning,
          middle,
          end,
        });

        if (Trigraphs.test(root)) {
          log.info("Exception identified", {
            type: PatternExceptions.Trigraph,
            value: root,
          });

          return `${beginning}${middle}${end}`;
        } else {
          return `${beginning};${middle}${this.#root({ root: end })}`;
        }
      }
    } else if (VV.test(tokens)) {
      log.info("Pattern match", {
        pattern: "VV",
      });

      const match = VV.exec(root);

      if (match?.groups) {
        const {
          groups: { beginning, end },
        } = match;

        log.debug("Match groups", {
          beginning,
          end,
        });

        if (Quadgraphs.test(root)) {
          log.info("Exception identified", {
            type: PatternExceptions.Quadgraph,
            value: root,
          });

          return `${beginning}${end}`;
        } else if (Digraphs.test(root)) {
          log.info("Exception identified", {
            type: PatternExceptions.Digraph.Digraph,
            value: root,
          });

          return `${beginning}${end}`;
        } else {
          log.info("No exception identified");

          return `${beginning};${this.#root({ root: end })}`;
        }
      }
    }

    log.info("No pattern match", {
      value: root,
    });

    return root;
  }

  #suffix({ cache, index, line }) {
    log.debug("Starting suffix parsing", {
      line,
    });

    const suffix = line.slice(index, line.length);

    if (Suffixes.has(suffix)) {
      log.info("New suffix identified", {
        index,
        suffix,
      });

      return [index, suffix];
    } else {
      log.debug("No new suffix identified", {
        index: cache.end,
        suffix: cache.suffix,
      });

      return [cache.end, cache.suffix];
    }
  }
}
