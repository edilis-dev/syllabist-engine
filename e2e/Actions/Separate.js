import { TextLineStream } from "@std/streams";

import { Separator } from "../../Separator.js";

export const defaults = {
  files: {
    input: "separate.txt",
    output: "separate.txt",
  },
};

export const separate = async ({
  files: { input = defaults.files.input, output = defaults.files.output },
} = defaults) => {
  const file = await Deno.open(input);
  const iter = file.readable
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(new TextLineStream());

  const data = await new Separator(iter).separate();

  await Deno.writeTextFile(output, data);
};
