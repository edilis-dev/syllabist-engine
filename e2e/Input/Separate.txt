aback
abandon
abandonment
abandoned
abated
abate
abatement
abet
abiotic
abiding
abide
ability
ablaze
able
ably
aboard
abode
abolish
abominable
abomination
abortion
abort
abound
about
above
abrasion
abrasive
abreast
abridged
abroad
abusive
abundance
abundant
abundantly
abuser
abusing
abuse
abused
abutting
aby
abysmal
abyss
acacia
academy
acetic
acetylene
achievable
achiever
achieving
achieve
achievement
achieved
acidity
acidic
acorn
acoustical
acoustic
acoustics
acre
acreage
acropolis
across
acrylic
acuity
acute
acutely
acyclic
adagio
adaptive
adaptor
adapt
adaptability
adaptable
adapter
adenosine
adieu
ado
adobe
adoption
adoptive
adopt
adopted
adopter
adorable
adore
adored
adorn
adrenaline
adrift
adultery
adult
adulthood
afar
aficionado
afield
afloat
afoot
aforementioned
aforesaid
