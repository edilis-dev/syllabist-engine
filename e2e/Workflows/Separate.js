import { assertEquals } from "@std/assert";
import { resolve } from "@std/path";

import { Standard } from "../../Constants.js";
import { Standardise } from "../../Format.js";
import * as Separator from "../Actions/Separate.js";
import { Cleanup } from "../Actions/Utility/Cleanup.js";
import { Setup } from "../Actions/Utility/Setup.js";

const { tmp } = await Setup();

await Separator.separate({
  files: {
    input: resolve(Deno.cwd(), "Input/Separate.txt"),
    output: resolve(tmp, Separator.defaults.files.output),
  },
});

const actual = await Deno.readTextFile(
  resolve(tmp, Separator.defaults.files.output),
);

console.log(resolve(tmp, Separator.defaults.files.output));

const expected = await Deno.readTextFile(
  resolve(Deno.cwd(), "Expected/Separate.txt"),
);

assertEquals(
  Standardise({ data: actual, type: Standard.Text }),
  Standardise({ data: expected, type: Standard.Text }),
);

await Cleanup({ tmp });
