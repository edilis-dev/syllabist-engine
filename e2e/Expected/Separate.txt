a;back
a;ban;don
a;ban;don;ment
a;ban;don;ed
a;bat;ed
a;bate
a;ba;te;ment
a;bet
a;bio;tic
a;bid;ing
a;bide
a;bi;lity
ab;la;ze
a;ble
ab;ly
a;board
a;bode
a;bo;lish
a;bomina;ble
a;bo;mi;na;tion
a;bor;ti;on
a;bort
a;bo;und
a;bout
a;bove
a;bra;sion
a;bra;sive
a;breast
a;bridg;ed
a;broad
a;bu;sive
a;bun;dan;ce
a;bun;dant
a;bun;dant;ly
a;bu;ser
a;bus;ing
a;buse
a;bus;ed
a;butt;ing
a;by
a;bys;mal
a;byss
a;ca;cia
a;ca;demy
a;ce;tic
a;cet;y;lene
a;chiev;a;ble
a;chiev;er
a;chiev;ing
a;chieve
a;chieve;ment
a;chieved
a;cid;i;ty
a;cid;ic
a;corn
a;cous;ti;cal
a;cous;tic
a;cous;tics
a;cre
a;cre;age
a;crop;o;lis
a;cross
a;cryl;ic
a;cu;i;ty
a;cute
a;cute;ly
a;cy;clic
a;da;gio
a;dap;tive
a;dap;tor
a;dapt
a;dapt;a;bil;i;ty
a;dapt;a;ble
a;dapt;er
a;den;o;sine
a;dieu
a;do
a;do;be
a;dop;tion
a;dop;tive
a;dopt
a;dopt;ed
a;dop;ter
a;dor;a;ble
a;dore
a;dored
a;dorn
a;dren;a;line
a;drift
a;dul;ter;y
a;dult
a;dult;hood
a;far
a;fi;cio;na;do
a;field
a;float
a;foot
a;fore;mentioned
a;fore;said
